### PR TITLE
Convert repository plugins to modules

### LINKED ISSUE #81652: Should we always install the main repository plugins?
This issue is for discussing whether we should always install the following plugins in Elasticsearch:

   - `repository-s3`
   - `repository-gcs`
   - `repository-azure`

Anyone running Elasticsearch one of the three main Cloud providers is likely using their object storage for backups, and other vendors also have S3-compatible system for performing backups in on-premise environments.

The upside of always bundling these plugins is that it relieves a piece of installation overhead, in particular making it easier to run the Docker image in a cloud environment.

The downside is that the distribution size will increase. There may be / are probably other downsides too.

We could:

   1. Convert these plugins to modules and completely bake them in. This makes upgrading the code a non-issue, since it just gets upgraded along with Elasticsearch.
   2. Keep them as plugins, which would give administrators the option to remove the plugins again if they wanted, but complicates the story around upgrades.

I would opt for (1) in order to avoid the complexity (and therefore bugs) associated with upgrades.

### PR DESCRIPTION
**Pull request by pugnascotia**
Closes #81652.

Convert the `repository-azure`, `repository-gcs` and `repository-s3` plugins into modules, so that they are always included in the Elasticsearch distribution. Also change plugin installation, removal and syncing so that attempting to add or remove these plugins still succeeds but is now a no-op.

### CODE REVIEW

File: distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/cli/InstallPluginAction.java [hunk 1]
```diff
@@ -219,6 +226,15 @@ public void execute(List<PluginDescriptor> plugins) throws Exception {
                     handleInstallXPack(buildFlavor());
                 }
 
+                if (PLUGINS_CONVERTED_TO_MODULES.contains(pluginId)) {
+                    // This deliberately does not throw an exception in order to avoid failing automation that relies on installing this
+                    // plugin during deployment.
+                    terminal.errorPrintln(
```
Comment (mark-vieira) [inline distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/cli/InstallPluginAction.java:232]: I wonder if additionally sending this to stderr will trip up any folks.

File: distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/cli/SyncPluginsAction.java [hunk 2]
```diff
@@ -115,7 +116,14 @@ public void execute() throws Exception {
     PluginChanges getPluginChanges(PluginsConfig pluginsConfig, Optional<PluginsConfig> cachedPluginsConfig) throws PluginSyncException {
         final List<PluginInfo> existingPlugins = getExistingPlugins();
 
-        final List<PluginDescriptor> pluginsThatShouldExist = pluginsConfig.getPlugins();
+        // For plugins that have migrated to modules, in order to help transition it's OK to still specify these plugins
```
Comment (mark-vieira) [inline distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/cli/SyncPluginsAction.java:119]: If that's the case should we even bother filtering here? Wouldn't it be better to just delegate on and get the warning message which would inform the user to remove these from their plugin config? As it is now we'd be silencing that message, no?

File: distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/cli/InstallPluginAction.java (hunk 1 above)
Comment (mark-vieira) [inline distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/cli/InstallPluginAction.java:232]: I wonder if there are additionally any concerns with sending this to stderrr? Would stdout similarly break stuff depending how folks are consuming the output?

File: docs/plugins/repository.asciidoc [hunk 3]
```diff
@@ -8,37 +8,20 @@ by distributed file systems:
 [discrete]
 ==== Core repository plugins
 
-The core repository plugins are:
-
-<<repository-s3,S3 Repository>>::
-
-The S3 repository plugin adds support for using S3 as a repository.
+NOTE: The S3, GCS and Azure repository plugins are now included in the {es}
```
Comment (mark-vieira) [inline docs/plugins/repository.asciidoc:11]: Since they're no longer "plugins" I'd reword this to say

> NOTE: Support for S3, GCS and Azure repository is now bundled in {es} by default

I'd also rename the section "Core repository plugins" to "Official repository plugins".

@elastic/es-docs thoughts?

File: modules/repository-s3/licenses/httpclient-NOTICE.txt [hunk 4]
```diff
@@ -1,5 +1,6 @@
 Apache HttpComponents Client
-Copyright 1999-2015 The Apache Software Foundation
+Copyright 1999-2016 The Apache Software Foundation
```
Comment (mark-vieira) [inline modules/repository-s3/licenses/httpclient-NOTICE.txt:2]: Did we update a dependency here?

**Review by mark-vieira — APPROVED** [review]
Couple of minor comments but otherwise LGTM. Looks like we have some minor packaging test fallout but I don't think there are any fundamental issues with the implementation.

File: modules/repository-s3/licenses/httpclient-NOTICE.txt (hunk 4 above)
Comment (pugnascotia) [reply modules/repository-s3/licenses/httpclient-NOTICE.txt:2]: No, but for some reason the copyright years were different between `repository-s3` and `repository-url`, so I went with the later year. The checksums of the actual jar files match.

File: distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/cli/SyncPluginsAction.java (hunk 2 above)
Comment (pugnascotia) [reply distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/cli/SyncPluginsAction.java:119]: The problem with delegating is that in this situation, the config file might say "repository-s3" should be installed, but actually we need to **remove** that plugin because it's now a module. So we need to change the "this plugin should be present" into a "this plugin should be removed". As a result, we don't reach the logic in `InstallPluginAction` that logs the warning about modularized plugins. However I realised that this implementation wasn't printing that warning at all when using a config file, so I've corrected that.

File: distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/cli/InstallPluginAction.java (hunk 1 above)
Comment (pugnascotia) [reply distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/cli/InstallPluginAction.java:232]: The thing is, the output from the CLI tools is completely unstructured. The only contract is the exit code of the tool, which is zero when installing or removing a modularized plugin. We can't reasonably be expected never to change the output of the CLI when it has no documented output format.

File: modules/build.gradle [hunk 5]
```diff
@@ -23,6 +23,11 @@ configure(subprojects.findAll { it.parent.path == project.path }) {
     throw new InvalidModelException("Modules cannot contain bin files")
   }
   if (project.file('src/main/config').exists()) {
-    throw new InvalidModelException("Modules cannot contain config files")
+    // The `config` directory is different, because we merge together log4j config files
```
Comment (pugnascotia) [inline modules/build.gradle:26]: @mark-vieira I don't know if this is the right thing to do here. The existing logic in `distribution/build.gradle` will pick up and combine any log4j config files in modules, yet we've banned all config files in modules. I could have just added the relevant logging config to the main config, but I prefer keeping the s3 config close to the relevant source code.

Comment (mark-vieira) [reply modules/build.gradle:26]: I think that is all OBE. You can see we don't enforce the same checks for x-pack/plugin (which are modules) and many of them contain config and bin files as well. I don't think we care about having different rules here for x-pack vs non-xpack. And in any case, if these rules were still applicable they should live in `InternalPluginBuildPlugin` not in the project build script.

In any case I think we can remove this. It makes no sense to forbid having config files, especially when we have [explicit handling for this](https://github.com/elastic/elasticsearch/blob/9a3422e1a6cf519e3fedce396784be2ef48dc7f9/distribution/build.gradle#L172-L177).

File: docs/reference/migration/migrate_8_0/plugin-changes.asciidoc [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+Specifically, the `elasticsearch-plugin` CLI tool will not fail if you
+attempt to install any of the above plugins, and will instead print a
+warning and skip the plugins. If any of these plugins are already
+installed, for example because you installed them when running an older
+version of {es}, then you can still remove them with
+`elasticsearch-plugin`. Attempting to remove them if they are not installed
+will succeed but print a warnings.
+
+If you run {es} using Docker and you are managing plugins using a
+<<manage-plugins-using-configuration-file,configuration file>>, then when
```
Comment (debadair) [inline docs/reference/migration/migrate_8_0/plugin-changes.asciidoc:43]:
```suggestion
{plugins}/manage-plugins-using-configuration-file.html[configuration file], then when
```

File: docs/reference/snapshot-restore/apis/get-repo-api.asciidoc [hunk 7]
```diff
@@ -97,13 +97,11 @@ URL repository. See <<snapshots-read-only-repository>>.
 
 More repository types are available through these official plugins:
 
-* {plugins}/repository-s3.html[repository-s3] for S3 repository support
+* <<repository-s3,repository-s3>> for S3 repository support
```
Comment (debadair) [inline docs/reference/snapshot-restore/apis/get-repo-api.asciidoc:100]:
```suggestion
* {plugins}/repository-s3[repository-s3] for S3 repository support
```

**Comment by debadair** [comment]
@elasticmachine retest this please

File: docs/reference/snapshot-restore/apis/get-repo-api.asciidoc [hunk 8]
```diff
@@ -100,10 +100,8 @@ More repository types are available through these official plugins:
 * {plugins}/repository-s3.html[repository-s3] for S3 repository support
```
Comment (debadair) [inline docs/reference/snapshot-restore/apis/get-repo-api.asciidoc:100]:
```suggestion
* <<repository-s3., repository-s3>> for S3 repository support
```

**Comment by debadair** [comment]
@elasticmachine retest this please

File: docs/reference/snapshot-restore/apis/get-repo-api.asciidoc [hunk 9]
```diff
@@ -97,13 +97,11 @@ URL repository. See <<snapshots-read-only-repository>>.
 
 More repository types are available through these official plugins:
 
-* {plugins}/repository-s3.html[repository-s3] for S3 repository support
+* <<repository-s3., repository-s3>> for S3 repository support
```
Comment (debadair) [inline docs/reference/snapshot-restore/apis/get-repo-api.asciidoc:100]:
```suggestion
* <<repository-s3, repository-s3>> for S3 repository support
```

**Comment by debadair** [comment]
@elasticmachine retest this please

**Comment by debadair** [comment]
The last broken links from the Kibana link service are fixed by elastic/kibana#121834. Once this PR is ready to go except for the broken links, it can be merged with the failing doc check and the Kibana PR can be retested & merged.

**Comment by pugnascotia** [comment]
@elasticmachine run elasticsearch-ci/cloud-deploy

**Comment by pugnascotia** [comment]
This PR requires changes from Cloud before we can merge.
